### PR TITLE
Track and Trace Protocol for SDK

### PR DESCRIPTION
**Pull request by RyanLassigBanks**
Adds track and trace protobuf definition that are compatible with Pike and Schema, as well as their corresponding native objects

### CODE REVIEW

**Comment by peterschwarz** [comment]
Those macros are cool. Can they be more broadly used for the other Native <-> proto conversions?

**Comment by RyanLassigBanks** [comment]
> Those macros are cool. Can they be more broadly used for the other Native <-> proto conversions?

Thank you! I spent a lot of time thinking about this. I had a couple of ideas of how to do it, but I couldn't quite come up with a syntax for the macro that felt good to use and wasn't overly complicated. I would have liked to spend more time thinking about it, but I felt moving the other TnT stories forward this sprint was a higher priority.

**Comment by RyanLassigBanks** [comment]
> What data type does this default result in?

The default for `PropertyDefinition` Would be this

```
PropertyDefinition {
    name: String::new(),
    data_type: DataType::Bytes,
    required: false,
    number_exponent: 0,
    enum_options: Vec::new(),
    struct_properties: Vec::new()
}
```

**Comment by RyanLassigBanks** [comment]
> This has a default implementation, which just returns `None`. Not really necessary until there are sources to return.

Fair. I remove the source impl then

File: sdk/protos/deprecated_track_and_trace_payload.proto [hunk 1]
```diff
@@ -0,0 +1,137 @@
+// Copyright 2017 Intel Corporation
```
Comment (agunde406) [inline sdk/protos/deprecated_track_and_trace_payload.proto]: typo in commit messsage implimentatin -> implementation

File: sdk/src/protocol/macros.rs [hunk 2]
```diff
@@ -0,0 +1,184 @@
+// Copyright 2019 Cargill Incorporated
```
Comment (agunde406) [inline sdk/src/protocol/macros.rs]: typo in commit message gerated -> generated

File: sdk/src/protocol/track_and_trace/payload.rs [hunk 3]
```diff
@@ -0,0 +1,471 @@
...
+        match actions {
+            TrackAndTracePayload_Action::CREATE_RECORD => Ok(Action::CreateRecord),
+            TrackAndTracePayload_Action::FINALIZE_RECORD => Ok(Action::FinalizeRecord),
+            TrackAndTracePayload_Action::UPDATE_PROPERTIES => Ok(Action::UpdateProperties),
+            TrackAndTracePayload_Action::CREATE_PROPOSAL => Ok(Action::CreateProposal),
+            TrackAndTracePayload_Action::ANSWER_PROPOSAL => Ok(Action::AnswerProposal),
+            TrackAndTracePayload_Action::REVOKE_REPORTER => Ok(Action::RevokeReporter),
+            TrackAndTracePayload_Action::UNSET_ACTION => {
+                Err(ProtoConversionError::InvalidTypeError(
+                    "Cannot convert SchemaPayload_Action with type unset.".to_string(),
```
Comment (agunde406) [inline sdk/src/protocol/track_and_trace/payload.rs]: SchemaPayload_Action -> TrackAndTracePayload_Action

File: sdk/src/protocol/track_and_trace/state.rs [hunk 4]
```diff
@@ -0,0 +1,611 @@
...
+        ));
+
+        Ok(proto)
+    }
+}
+
+impl_from_bytes!(RecordList, track_and_trace_state::RecordList);
+impl_into_bytes!(RecordList);
+impl_into_proto!(RecordList, track_and_trace_state::RecordList);
+impl_into_native!(RecordList, track_and_trace_state::RecordList);
```
Comment (agunde406) [inline sdk/src/protocol/track_and_trace/state.rs]: Please add unit tests.

File: sdk/src/protocol/track_and_trace/payload.rs [hunk 5]
```diff
@@ -0,0 +1,471 @@
...
+);
+impl_into_bytes!(TrackAndTracePayload);
+impl_into_proto!(
+    TrackAndTracePayload,
+    track_and_trace_payload::TrackAndTracePayload
+);
+impl_into_native!(
+    TrackAndTracePayload,
+    track_and_trace_payload::TrackAndTracePayload
+);
```
Comment (agunde406) [inline sdk/src/protocol/track_and_trace/payload.rs]: Please add unit tests

File: sdk/src/protocol/macros.rs [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+///         self.record_id = Some(value);
+///         self
+///     }
+///
+///     pub fn name(mut self, value: &str) -> Self {
+///         self.name = Some(value);
+///         self
+///     }
+///
+///     pub fn build(self) -> Result<FianlizeRecordAction, BuilderError> {
```
Comment (vaporos) [inline sdk/src/protocol/macros.rs]: Typo on Finalized

File: sdk/src/protocol/schema/state.rs [hunk 7]
```diff
@@ -110,7 +117,7 @@ impl IntoProto<protos::schema_state::LatLong> for LatLong {}
 impl IntoNative<LatLong> for protos::schema_state::LatLong {}
 
 /// Native implementation of PropertyDefinition
-#[derive(Debug, Clone, PartialEq)]
+#[derive(Default, Debug, Clone, PartialEq)]
```
Comment (peterschwarz) [inline sdk/src/protocol/schema/state.rs]: I'm not sure Default makes sense on this structure - I don't think empty name is allowed.

File: sdk/src/protocol/schema/state.rs [hunk 8]
```diff
@@ -677,7 +684,7 @@ impl SchemaListBuilder {
 }
 
 /// Native implementation of PropertyValue
-#[derive(Debug, Clone, PartialEq)]
+#[derive(Default, Debug, Clone, PartialEq)]
```
Comment (peterschwarz) [inline sdk/src/protocol/schema/state.rs]: I don't think `Default` makes sense here either.

File: sdk/src/protocol/track_and_trace/state.rs [hunk 9]
```diff
@@ -0,0 +1,1886 @@
...
+    pub fn with_fixed(mut self, value: bool) -> Self {
+        self.fixed = Some(value);
+        self
+    }
+    pub fn with_unit(mut self, value: String) -> Self {
+        self.unit = Some(value);
+        self
+    }
+    pub fn build(self) -> Result<Property, BuilderError> {
+        let name = self.name;
```
Comment (peterschwarz) [inline sdk/src/protocol/track_and_trace/state.rs]: Not sure I agree with all this name-shadowing in these build methods.

File: contracts/track_and_trace/src/handler.rs [hunk 10]
```diff
@@ -32,20 +32,32 @@ cfg_if! {
     }
 }
 
-use grid_sdk::protos::track_and_trace_agent::TrackAndTraceAgent as Agent;
-use grid_sdk::protos::track_and_trace_payload::CreateTrackAndTraceAgentAction as CreateAgentAction;
-use grid_sdk::protos::track_and_trace_payload::{
-    AnswerProposalAction, AnswerProposalAction_Response, CreateProposalAction, CreateRecordAction,
-    CreateRecordTypeAction, FinalizeRecordAction, RevokeReporterAction, UpdatePropertiesAction,
+use grid_sdk::protos::deprecated_track_and_trace_payload::DeprecatedCreateTrackAndTraceAgentAction as CreateAgentAction;
```
Comment (peterschwarz) [inline contracts/track_and_trace/src/handler.rs]: Typo in commit message: depricated -> deprecated

**Pull request merged**